dbms/src/Flash/Pipeline/Pipeline.cpp

```diff
@@ -0,0 +1,148 @@
+// Copyright 2023 PingCAP, Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Common/FmtUtils.h>
+#include <Flash/Coprocessor/FineGrainedShuffle.h>
+#include <Flash/Executor/PipelineExecutorStatus.h>
+#include <Flash/Pipeline/Exec/PipelineExecBuilder.h>
+#include <Flash/Pipeline/Pipeline.h>
+#include <Flash/Pipeline/Schedule/Event/Event.h>
+#include <Flash/Pipeline/Schedule/Event/PlainPipelineEvent.h>
+#include <Flash/Planner/PhysicalPlanNode.h>
+#include <Flash/Planner/plans/PhysicalGetResultSink.h>
+#include <Flash/Statistics/traverseExecutors.h>
+#include <tipb/select.pb.h>
+
+namespace DB
+{
+namespace
+{
+FmtBuffer & addPrefix(FmtBuffer & buffer, size_t level)
+{
+    return buffer.append(String(level, ' '));
+}
+} // namespace
+
+void Pipeline::addPlanNode(const PhysicalPlanNodePtr & plan_node)
+{
+    plan_nodes.push_back(plan_node);
+}
+
+void Pipeline::addDependency(const PipelinePtr & dependency)
+{
+    dependencies.emplace_back(dependency);
+}
+
+void Pipeline::toSelfString(FmtBuffer & buffer, size_t level) const
+{
+    size_t prefix_size = 2 * level;
+    addPrefix(buffer, prefix_size).append("pipeline:\n");
+    ++prefix_size;
+    for (const auto & plan_node : plan_nodes)
+        addPrefix(buffer, prefix_size).append(plan_node->toString()).append("\n");
+}
+
+void Pipeline::toTreeString(FmtBuffer & buffer, size_t level) const
+{
+    toSelfString(buffer, level);
+    ++level;
+    for (const auto & dependency : dependencies)
+    {
+        auto dependency_ptr = dependency.lock();
+        assert(dependency_ptr);
+        dependency_ptr->toTreeString(buffer, level);
+    }
+}
+
+void Pipeline::addGetResultSink(ResultHandler result_handler)
+{
+    assert(!plan_nodes.empty());
+    auto get_result_sink = PhysicalGetResultSink::build(result_handler, plan_nodes.front());
+    plan_nodes.push_front(get_result_sink);
+    get_result_sink->detach();
+}
+
+PipelineExecGroup Pipeline::toExecGroup(Context & context, size_t concurrency)
```

Comment:
I'm raising a question that may concern the coming `FineGrainedPipelineEvent`: Now `PlainPipelineEvent` delegates the `PipelineExec` generation to `Pipeline`. But I imagine there could be fine-grained-specific logic to the `PipelineExec` for `FineGrainedPipelineEvent`, e.g. a specified partition id. How does this method handle this situation?

Reply:
Maybe I will create a new function `Pipeline::toFineGrainedExec(Context & context, size_t partition_id)`.

Reply:
Though not a big problem, this doesn't look very elegant: You have one `Pipeline` impl, one `PipelineExec` impl, but two `PipelineEvent` impls. The difference between "plain" `PipelineExec` and "fine-grained" `PipelineExec` should be hided by the two `PipelineEvent`s, not exposed in `Pipeline` level.

Reply:
The logic for plain and fine grained shuffle is written in `PhysicalPlanNode::buildBlockInputStream` with if else, here the logic of the two is separated.